Optimize lsra for MinOpts

kunalspathak opened the pull request:
- Created separate methods *Minimal() and moved the `enregisterLocalVars=false` specific code in it
- Only run `REG_ORDER` and `REG_NUM` heuristics when `enregisterLocalVars=false`

kunalspathak committed:
Optimize resetAllRegistersState()

kunalspathak committed:
Add assert about refType for minopts

kunalspathak committed:
Templatize processBlockEndAllocation()

kunalspathak committed:
Add resetMinOpts()

kunalspathak committed:
Add selectMinOpts()

kunalspathak committed:
Add calculateUnassignedSet()

kunalspathak committed:
Call spillCostMinOpts() for minopts

kunalspathak committed:
Add enregisterLocalVars check in setIntervalAsSpilled()

kunalspathak committed:
optimize few paths in resolveRegisters()

kunalspathak committed:
optimize few paths in allocateRegisters()

kunalspathak committed:
Adjust RegisterSelection creation

kunalspathak committed:
Add allocateRegistersForMinOpts()

kunalspathak committed:
jit format

kunalspathak commented:
Need to find out what is going on with arm64, but initial results are very motivating!

![image](https://github.com/dotnet/runtime/assets/12488060/76e9707c-6dce-4598-b890-24fd62c32605)

![image](https://github.com/dotnet/runtime/assets/12488060/6fbf2375-2690-437a-a592-09443c8c51c4)

kunalspathak committed:
Update the assert for processBlockStartLocations()

kunalspathak commented:
I tried to investigate the windows-arm64 TP regression

<img width="465" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/e2d75f3b-33fb-4a5b-abc6-cea38ff2adab">

It turns out that the majority of slowness is attributed to 2 method contexts of `benchmarks.run.windows.arm64.checked.mch` : `2999` and `15658` and the reason being the majority of gains from register allocation are compensated by the equal work that emitter has to do to produce more instructions. Below is the `pin` difference for just 2 method contexts.

```
Base: 414060, Diff: 433023, +4.5798%

??$allocateReg@$0A@@LinearScan@@AEAA?AW4_regNumber_enum@@PEAVInterval@@PEAVRefPosition@@@Z : 24193  : +942.46% : 13.83% : +5.8429%
??$allocateRegistersForMinOpt@$0A@@LinearScan@@QEAAXXZ                                     : 22858  : NA       : 13.07% : +5.5205%
?emitOutputInstr@emitter@@IEAA_KPEAUinsGroup@@PEAUinstrDesc@1@PEAPEAE@Z                    : 8062   : +61.79%  : 4.61%  : +1.9471%
?lvaFrameAddress@Compiler@@QEAAHHPEA_N@Z                                                   : 6491   : +575.44% : 3.71%  : +1.5676%
?emitAllocAnyInstr@emitter@@AEAAPEAX_KW4emitAttr@@@Z                                       : 2664   : +45.73%  : 1.52%  : +0.6434%
?emitIns_S_R@emitter@@QEAAXW4instruction@@W4emitAttr@@W4_regNumber_enum@@HH@Z              : 2568   : +143.87% : 1.47%  : +0.6202%
?rsSpillTree@RegSet@@AEAAXW4_regNumber_enum@@PEAUGenTree@@I@Z                              : 2544   : NA       : 1.45%  : +0.6144%
?emitIns_R_S@emitter@@QEAAXW4instruction@@W4emitAttr@@W4_regNumber_enum@@HH@Z              : 2391   : +127.05% : 1.37%  : +0.5775%
?IsMultiRegNode@GenTree@@QEBA_NXZ                                                          : 1988   : +177.03% : 1.14%  : +0.4801%
?emitInsCode@emitter@@AEAAIW4instruction@@W4insFormat@1@@Z                                 : 1496   : +65.99%  : 0.86%  : +0.3613%
?idCodeSize@instrDesc@emitter@@QEBAIXZ                                                     : 1290   : +45.94%  : 0.74%  : +0.3115%
?updateMaxSpill@LinearScan@@QEAAXPEAVRefPosition@@@Z                                       : 1267   : +54.61%  : 0.72%  : +0.3060%
?resetAllRegistersState@LinearScan@@AEAAXXZ                                                : 1255   : NA       : 0.72%  : +0.3031%
?unassignPhysReg@LinearScan@@AEAAXPEAVRegRecord@@PEAVRefPosition@@@Z                       : 1232   : +47.26%  : 0.70%  : +0.2975%
memset                                                                                     : 1218   : +44.71%  : 0.70%  : +0.2942%
?initLclVarAddr@emitLclVarAddr@@QEAAXHI@Z                                                  : 1075   : +139.97% : 0.61%  : +0.2596%
?emitSizeOfInsDsc@emitter@@AEBA_KPEAUinstrDesc@1@@Z                                        : 1056   : +106.88% : 0.60%  : +0.2550%
?genUnspillRegIfNeeded@CodeGen@@IEAAXPEAUGenTree@@@Z                                       : 996    : +101.01% : 0.57%  : +0.2405%
?insertCopyOrReload@LinearScan@@QEAAXPEAUBasicBlock@@PEAUGenTree@@IPEAVRefPosition@@@Z     : 935    : NA       : 0.53%  : +0.2258%
?emitEndCodeGen@emitter@@QEAAIPEAVCompiler@@_N11IPEAI2PEAPEAX33@Z                          : 903    : +26.23%  : 0.52%  : +0.2181%
?updateAssignedInterval@LinearScan@@AEAAXPEAVRegRecord@@PEAVInterval@@@Z                   : 777    : +35.41%  : 0.44%  : +0.1877%
?appendToCurIG@emitter@@IEAAXPEAUinstrDesc@1@@Z                                            : 731    : +52.18%  : 0.42%  : +0.1765%
?rsUnspillInPlace@RegSet@@AEAAPEAVTempDsc@@PEAUGenTree@@W4_regNumber_enum@@I@Z             : 711    : NA       : 0.41%  : +0.1717%
?assignPhysReg@LinearScan@@AEAAXPEAVRegRecord@@PEAVInterval@@@Z                            : 693    : +36.21%  : 0.40%  : +0.1674%
?emitFindOffset@emitter@@IEBAIPEBUinsGroup@@I@Z                                            : 672    : +72.41%  : 0.38%  : +0.1623%
?insEncodeDatasizeLS@emitter@@CAIIW4emitAttr@@@Z                                           : 598    : +120.56% : 0.34%  : +0.1444%
?tmpRlsTemp@RegSet@@QEAAXPEAVTempDsc@@@Z                                                   : 588    : NA       : 0.34%  : +0.1420%
?TryGetUse@Range@LIR@@QEAA_NPEAUGenTree@@PEAVUse@2@@Z                                      : 539    : NA       : 0.31%  : +0.1302%
?genCodeForTreeNode@CodeGen@@IEAAXPEAUGenTree@@@Z                                          : 451    : +9.59%   : 0.26%  : +0.1089%
?emitGCregDeadUpd@emitter@@QEAAXW4_regNumber_enum@@PEAE@Z                                  : 402    : +33.67%  : 0.23%  : +0.0971%
??$resolveRegisters@$0A@@LinearScan@@QEAAXXZ                                               : 396    : +4.76%   : 0.23%  : +0.0956%
?allocateMemory@ArenaAllocator@@QEAAPEAX_K@Z                                               : 368    : +3.28%   : 0.21%  : +0.0889%
?emitNewInstrCns@emitter@@AEAAPEAUinstrDesc@1@W4emitAttr@@_J@Z                             : 357    : +87.50%  : 0.20%  : +0.0862%
?tmpSlot@RegSet@@CAII@Z                                                                    : 286    : NA       : 0.16%  : +0.0691%
?setIntervalAsSpilled@LinearScan@@AEAAXPEAVInterval@@@Z                                    : 252    : NA       : 0.14%  : +0.0609%
?gcMakeRegPtrTable@GCInfo@@QEAAXPEAVGcInfoEncoder@@IIW4MakeRegPtrMode@1@PEAI@Z             : 244    : +10.95%  : 0.14%  : +0.0589%
?ins_Load@CodeGenInterface@@QEAA?AW4instruction@@W4var_types@@_N@Z                         : 237    : +179.55% : 0.14%  : +0.0572%
?lvaAllocateTemps@Compiler@@QEAAHH_N@Z                                                     : 221    : +153.47% : 0.13%  : +0.0534%
?TryGetUse@GenTree@@QEAA_NPEAU1@PEAPEAPEAU1@@Z                                             : 212    : NA       : 0.12%  : +0.0512%
?getWeight@LinearScan@@AEAANPEAVRefPosition@@@Z                                            : 189    : +10.41%  : 0.11%  : +0.0456%
?recordMaxSpill@LinearScan@@QEAAXXZ                                                        : 182    : +55.83%  : 0.10%  : +0.0440%
?UpdateLifeVar@?$TreeLifeUpdater@$00@@AEAAXPEAUGenTree@@PEAUGenTreeLclVarCommon@@@Z        : -175   : -21.74%  : 0.10%  : -0.0423%
?getMatchingConstants@LinearScan@@AEAA_K_KPEAVInterval@@PEAVRefPosition@@@Z                : -180   : -100.00% : 0.10%  : -0.0435%
?processBlockEndAllocation@LinearScan@@AEAAXPEAUBasicBlock@@@Z                             : -1896  : -100.00% : 1.08%  : -0.4579%
??$allocateRegisters@$0A@@LinearScan@@QEAAXXZ                                              : -30030 : -100.00% : 17.17% : -7.2526%
??$select@$0A@@RegisterSelection@LinearScan@@QEAA_KPEAVInterval@@PEAVRefPosition@@@Z       : -45417 : -100.00% : 25.97% : -10.9687%

```

If I exclude those 2 contexts from the `benchmarks.run.windows` collection, then locally I do see TP gains that matches for other platforms:

<img width="639" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/e0a112ad-0b61-4be1-8fbb-d670481bf2da">

[2999 diff](https://www.diffchecker.com/y3QqdE7m/) and [15658 diff](https://www.diffchecker.com/MUmgQPfU/)

kunalspathak committed:
Fix bug by using freeCandidates

kunalspathak commented:
There was a bug where I was almost always assigning busy register. That was having us generate lot of spilling and emit spilling code increase the TP. I fixed that in https://github.com/dotnet/runtime/pull/96386/commits/7f29c4e7048fcfb41ed9d009dfa93c73b7eb69d9. The new numbers are even better, especially on Arm64, we see almost 13% TP improvements. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=513754&view=ms.vss-build-web.run-extensions-tab):

![image](https://github.com/dotnet/runtime/assets/12488060/909416f9-8fa0-4012-b9bc-38c448165d85)

![image](https://github.com/dotnet/runtime/assets/12488060/aca072ec-6288-47af-b51f-516429233901)

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra

kunalspathak committed:
Set candidates only if freeCandidates are NONE

kunalspathak committed:
Remove unnecessary setting of foundRegBit

kunalspathak committed:
Remove unnecessary checks for `freeCandidates == RBM_NONE`

kunalspathak committed:
Remove unnecessary checks for `freeCandidates == RBM_NONE` --amend

kunalspathak committed:
wip

kunalspathak committed:
add allocateRegForMinOpts()

kunalspathak committed:
Remove hasConsecutiveRegister template from allocateRegistersForMinOpt()

kunalspathak committed:
Add template parameter for assignCopyReg() and fix the uses

kunalspathak committed:
fix calls to processBlockEndAllocation()

kunalspathak committed:
fix bug about initializing vectorIter

kunalspathak committed:
Add assignCopyRegForMinOpts

kunalspathak committed:
jit format

kunalspathak committed:
rename from *MinOpts to *Minimal

kunalspathak commented:
/azp run runtime-coreclr jitstressregs, runtime-coreclr libraries-jitstressregs, runtime-coreclr libraries-jitstress2-jitstressregs, runtime-coreclr jitstressregs-x86, runtime-coreclr jitstress2-jitstressregs

kunalspathak committed:
Remove unneeded method

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra

kunalspathak committed:
jit format

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra

kunalspathak commented:
/azp run runtime-coreclr jitstressregs, runtime-coreclr libraries-jitstressregs, runtime-coreclr libraries-jitstress2-jitstressregs, runtime-coreclr jitstressregs-x86, runtime-coreclr jitstress2-jitstressregs

kunalspathak commented:
All failures are known and https://github.com/dotnet/runtime/issues/93194

jakobbotsch commented:
Why are there regressions in FullOpts?

kunalspathak commented:
> Why are there regressions in FullOpts?

I assume you are asking about asmdiff and not tpdiff? I will double check.

jakobbotsch commented:
> > Why are there regressions in FullOpts?
> 
> I assume you are asking about asmdiff and not tpdiff? I will double check.

Yeah, the tpdiffs look amazing, but I assumed the FullOpts diffs are unexpected.

jakobbotsch commented:
> right, that's what I was asking offline sometime back. A method can be compiled in Fullopts, but we might have `enregisterLocalVar=false` and those methods will be affected by this chance. I can find 1 or 2 examples of that.

I see a number of cases in the example diffs that are FullOpts cases and where locals in the locals table are enregistered. Wouldn't we expect locals not to be enregistered then?
Example:
![image](https://github.com/dotnet/runtime/assets/7887810/122cec75-cb93-45a7-ab24-8cdac61ac06d)

kunalspathak commented:
> I see a number of cases in the example diffs that are FullOpts cases and where locals in the locals table are enregistered. Wouldn't we expect locals not to be enregistered then?

Thanks for pointing out the example. Yes, let me double check what is going on with them.

BruceForstall commented:
The asmdiffs are huge and varied: huge improvements on some platforms and huge regressions on others (both for MinOpts).

And, e.g., huge regressions in FullOpts on win-x86, where it looks like FullOpts code is getting MinOpts reg allocation?

kunalspathak committed:
Add back short-circuit code

kunalspathak commented:
Some diffs in FullOpts is because we set `enregisterLocalVars=false` if they are no tracking variables. Others were present because I removed the short-circuiting code under `#ifdef DEBUG` that would execute the selection heuristics, which was earlier skipped. This changed the `candidates` mask and hence saw massive diffs. In `Release`, things are as expected. Fixed to put back the short-circuit code in DEBUG.

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=531533&view=ms.vss-build-web.run-extensions-tab)

Still pretty sizable regressions on (most) FullOpts.

Why does MinOpts have just drastic differences, namely, ~150K regression on linux-arm64 but 464K improvement on linux-x64, and only 1K regression on win-x64?

kunalspathak commented:
Yes, so here is the breakdown:

FullOpts:
1. Blue - All x64 platforms is showing around 17K+ regression
2. Green - All arm64 platforms is showing around 32K+ regression (acceptable because of arm64 larger code size)

MinOpts:
1. Yellow - All arm64 platforms MinOpts is showing around 145K+ regression
2. Purple - Linux/x64 platform MinOpts is showing 464K+ improvement


<img width="438" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/a2ad3393-956d-4bbd-9b84-b9dc8e8d391d">

Looking in more detail, I checked what % of contents are affected in each of the configuration. For x64, there are around 28% (both windows/linux) and on arm64, there are around 15% (both windows/linux), so we at least see that the contexts affected are same for a particular architecture.

| OS | arch | total context | diff context | % diff |
|---|---|---|---|---|
| windows | x64 | 2384370 | 666648 | 28% |
| linux | x64 | 2546536 | 716232 | 28% |
| windows | arm64 | 2329306 | 366025 | 16% |
| linux | arm64 | 2495514 | 385512 | 15% |
| windows | x86 | 2244188 | 505186 | 23% |
| linux | arm | 2239583 | 671422 | 30% |


#### Linux/x64 analysis

Values are mostly in `rax` (because of `REG_ORDER`), I see that instructions are more compact with it, Here is an example from `benchmarks.run.linux` collection

```diff
-  400FB67F4C          movzx    rdi, byte  ptr [rdi+0x4C]
+ 0FB6404C             movzx    rax, byte  ptr [rax+0x4C]
```

Some other places, since value is mostly in `rax`, I see `cwde` instruction is used instead of `movsx` which is smaller in size:

<img width="1542" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/1c6f9890-758c-4ad1-b633-ffb97b7cc351">

<img width="1427" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/5ef652c7-aa71-47e4-a2cd-2bdb6b41f01b">

That could justify the improvements on linux, but if you want I can dig further.

jakobbotsch commented:
IMO, since the change is about MinOpts, it should be no-diff for FullOpts. How does the change end up affecting codegen for FullOpts in the cases where we have no tracking variables? Isn't that unexpected? And if it's expected, can the new logic be made to run only if `OptimizationDisabled() && enregisterLocalVar` or something?

kunalspathak commented:
> IMO, since the change is about MinOpts, it should be no-diff for FullOpts. 

This change is not just for MinOpts, but when we decide that enregistering of local vars is not needed. This is fairly common in MinOpts/Tier0/`MethodImplOptions.NoOptimization` but in FullOpts, when there are no tracking variables. 

> How does the change end up affecting codegen for FullOpts in the cases where we have no tracking variables? Isn't that unexpected? 

We turn off `enregisterLocalVars` if we don't have any tracking variable and hence this change is affecting exactly and only those scenarios for FullOpts.

https://github.com/dotnet/runtime/blob/62d33ee48d57feba67b261b55db666bdc202b1c1/src/coreclr/jit/lsra.cpp#L1353-L1358

> And if it's expected, can the new logic be made to run only if `OptimizationDisabled() && !enregisterLocalVar` or something? Edit: It makes sense to me that we'd still want the optimized heuristics for LIR temps, even when we don't have any locals to enregister, so I think we should base the use of the fast and less optimal heuristics around OptimizationDisabled() instead.

I am not sure I follow what you are suggesting.

jakobbotsch commented:
> We turn off enregisterLocalVars if we don't have any tracking variable and hence this change is affecting exactly and only those scenarios for FullOpts.

The end result seems to be that we are regressing CQ in FullOpts for a throughput optimization. That doesn't seem right.

`enregisterLocalVars` does not seem like the right check to me. Sure if there are no local vars then there are also no local vars to enregister, even in FullOpts, but that doesn't mean there are no other intervals that should get good register allocation.

> I am not sure I follow what you are suggesting.

Call `allocateRegistersMinimal()` only if `opts.OptimizationDisabled()`. Optionally assert that `enregisterLocalVars` is false in that case. Keep allocating all intervals with the full set of heuristics in FullOpts.

BruceForstall commented:
If @jakobbotsch suggestion restricts the change the MinOpts, then this change could be only about MinOpts CQ + TP. @kunalspathak if you think a version of this change applies to FullOpts, then perhaps that could be made as a follow-up, to isolate the conversation about the FullOpts impact to that PR.

kunalspathak commented:
> Call `allocateRegistersMinimal()` only if `opts.OptimizationDisabled()`. Optionally assert that `enregisterLocalVars` is false in that case. Keep allocating all intervals with the full set of heuristics in FullOpts.

ah i see. Yes, I was trying out similar change locally to eliminate the full opt impact.

kunalspathak committed:
Quick LSRA only if optimization is disabled

kunalspathak commented:
superpmi failures are: OSError: [Errno 28] No space left on device

/azp run runtime-coreclr superpmi-replay

kunalspathak commented:
/azp run runtime-coreclr superpmi-replay

kunalspathak commented:
![image](https://github.com/dotnet/runtime/assets/12488060/133d59e9-a7a3-4253-b8ba-2491d3beec03)
![image](https://github.com/dotnet/runtime/assets/12488060/074b427c-d162-42b2-842e-d744fe6b549d)
![image](https://github.com/dotnet/runtime/assets/12488060/7a4c1d65-6741-43f7-881c-0f0690f5adb8)

kunalspathak commented:
## Assembly diffs

### Assembly diffs for linux/arm64 ran on windows/x64
Diffs are based on ${\color{orange}2,491,707}$ contexts (${\color{orange}1,001,189}$ MinOpts, ${\color{orange}1,490,518}$ FullOpts).

${\color{blue}MISSED}$ contexts: ${\color{blue}3,807 (0.15\\%)}$


<details>
<summary>Overall (${\color{red}+148,204}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.linux.arm64.checked.mch|14,270,080|+0|
|benchmarks.run_pgo.linux.arm64.checked.mch|83,058,256|${\color{red}+1,980}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|21,983,044|${\color{red}+588}$|
|coreclr_tests.run.linux.arm64.checked.mch|510,170,492|${\color{red}+144,156}$|
|libraries.crossgen2.linux.arm64.checked.mch|55,689,604|+0|
|libraries.pmi.linux.arm64.checked.mch|75,931,904|+0|
|libraries_tests.run.linux.arm64.Release.mch|379,668,404|${\color{red}+1,372}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|162,138,568|${\color{red}+108}$|
|realworld.run.linux.arm64.checked.mch|15,784,696|+0|


</div></details>

<details>
<summary>MinOpts (${\color{red}+148,204}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.linux.arm64.checked.mch|218,508|+0|
|benchmarks.run_pgo.linux.arm64.checked.mch|26,552,600|${\color{red}+1,980}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|17,400,224|${\color{red}+588}$|
|coreclr_tests.run.linux.arm64.checked.mch|349,264,408|${\color{red}+144,156}$|
|libraries.crossgen2.linux.arm64.checked.mch|1,636|+0|
|libraries.pmi.linux.arm64.checked.mch|119,984|+0|
|libraries_tests.run.linux.arm64.Release.mch|214,980,496|${\color{red}+1,372}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|13,480,568|${\color{red}+108}$|
|realworld.run.linux.arm64.checked.mch|575,816|+0|


</div></details>

### Assembly diffs for linux/x64 ran on windows/x64
Diffs are based on ${\color{orange}2,542,496}$ contexts (${\color{orange}1,013,312}$ MinOpts, ${\color{orange}1,529,184}$ FullOpts).

${\color{blue}MISSED}$ contexts: ${\color{blue}4,042 (0.16\\%)}$


<details>
<summary>Overall (${\color{green}-464,740}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.linux.x64.checked.mch|13,063,008|${\color{green}-5}$|
|benchmarks.run_pgo.linux.x64.checked.mch|64,981,453|${\color{green}-15,309}$|
|benchmarks.run_tiered.linux.x64.checked.mch|17,221,305|${\color{green}-11,296}$|
|coreclr_tests.run.linux.x64.checked.mch|439,635,202|${\color{green}-246,288}$|
|libraries.crossgen2.linux.x64.checked.mch|38,636,779|${\color{green}-4}$|
|libraries.pmi.linux.x64.checked.mch|59,927,414|${\color{green}-13}$|
|libraries_tests.run.linux.x64.Release.mch|328,845,908|${\color{green}-136,481}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|129,732,524|${\color{green}-54,318}$|
|realworld.run.linux.x64.checked.mch|13,174,701|${\color{green}-1,026}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|4,191,403|+0|


</div></details>

<details>
<summary>MinOpts (${\color{green}-464,740}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.linux.x64.checked.mch|230,051|${\color{green}-5}$|
|benchmarks.run_pgo.linux.x64.checked.mch|18,706,882|${\color{green}-15,309}$|
|benchmarks.run_tiered.linux.x64.checked.mch|13,644,734|${\color{green}-11,296}$|
|coreclr_tests.run.linux.x64.checked.mch|310,584,524|${\color{green}-246,288}$|
|libraries.crossgen2.linux.x64.checked.mch|1,202|${\color{green}-4}$|
|libraries.pmi.linux.x64.checked.mch|112,870|${\color{green}-13}$|
|libraries_tests.run.linux.x64.Release.mch|183,459,665|${\color{green}-136,481}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|10,657,089|${\color{green}-54,318}$|
|realworld.run.linux.x64.checked.mch|387,256|${\color{green}-1,026}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|911|+0|


</div></details>

### Assembly diffs for osx/arm64 ran on windows/x64
Diffs are based on ${\color{orange}2,257,766}$ contexts (${\color{orange}928,286}$ MinOpts, ${\color{orange}1,329,480}$ FullOpts).

${\color{blue}MISSED}$ contexts: ${\color{blue}2,387 (0.11\\%)}$


<details>
<summary>Overall (${\color{red}+150,496}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run_pgo.osx.arm64.checked.mch|34,138,564|${\color{red}+1,480}$|
|benchmarks.run_tiered.osx.arm64.checked.mch|15,442,448|${\color{red}+316}$|
|coreclr_tests.run.osx.arm64.checked.mch|486,102,776|${\color{red}+148,116}$|
|libraries.crossgen2.osx.arm64.checked.mch|55,570,716|+0|
|libraries.pmi.osx.arm64.checked.mch|79,907,992|+0|
|libraries_tests.run.osx.arm64.Release.mch|314,762,740|${\color{red}+476}$|
|libraries_tests_no_tiered_compilation.run.osx.arm64.Release.mch|159,436,088|${\color{red}+108}$|
|realworld.run.osx.arm64.checked.mch|15,065,616|+0|


</div></details>

<details>
<summary>MinOpts (${\color{red}+150,496}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run_pgo.osx.arm64.checked.mch|16,484,956|${\color{red}+1,480}$|
|benchmarks.run_tiered.osx.arm64.checked.mch|11,499,424|${\color{red}+316}$|
|coreclr_tests.run.osx.arm64.checked.mch|332,741,972|${\color{red}+148,116}$|
|libraries.crossgen2.osx.arm64.checked.mch|1,628|+0|
|libraries.pmi.osx.arm64.checked.mch|121,128|+0|
|libraries_tests.run.osx.arm64.Release.mch|203,635,640|${\color{red}+476}$|
|libraries_tests_no_tiered_compilation.run.osx.arm64.Release.mch|13,136,596|${\color{red}+108}$|
|realworld.run.osx.arm64.checked.mch|568,396|+0|


</div></details>

### Assembly diffs for windows/arm64 ran on windows/x64
Diffs are based on ${\color{orange}2,326,811}$ contexts (${\color{orange}934,686}$ MinOpts, ${\color{orange}1,392,125}$ FullOpts).

${\color{blue}MISSED}$ contexts: ${\color{blue}2,495 (0.11\\%)}$


<details>
<summary>Overall (${\color{red}+148,732}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run_pgo.windows.arm64.checked.mch|45,619,340|${\color{red}+1,376}$|
|benchmarks.run_tiered.windows.arm64.checked.mch|15,264,644|${\color{red}+368}$|
|coreclr_tests.run.windows.arm64.checked.mch|496,117,708|${\color{red}+144,188}$|
|libraries.crossgen2.windows.arm64.checked.mch|58,913,600|+0|
|libraries.pmi.windows.arm64.checked.mch|79,525,232|+0|
|libraries_tests.run.windows.arm64.Release.mch|319,612,456|${\color{red}+2,688}$|
|libraries_tests_no_tiered_compilation.run.windows.arm64.Release.mch|169,012,984|${\color{red}+112}$|
|realworld.run.windows.arm64.checked.mch|15,917,176|+0|


</div></details>

<details>
<summary>MinOpts (${\color{red}+148,732}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run_pgo.windows.arm64.checked.mch|16,205,148|${\color{red}+1,376}$|
|benchmarks.run_tiered.windows.arm64.checked.mch|11,172,336|${\color{red}+368}$|
|coreclr_tests.run.windows.arm64.checked.mch|339,719,232|${\color{red}+144,188}$|
|libraries.crossgen2.windows.arm64.checked.mch|1,636|+0|
|libraries.pmi.windows.arm64.checked.mch|119,984|+0|
|libraries_tests.run.windows.arm64.Release.mch|203,904,856|${\color{red}+2,688}$|
|libraries_tests_no_tiered_compilation.run.windows.arm64.Release.mch|13,136,180|${\color{red}+112}$|
|realworld.run.windows.arm64.checked.mch|568,424|+0|


</div></details>

### Assembly diffs for windows/x64 ran on windows/x64
Diffs are based on ${\color{orange}2,495,240}$ contexts (${\color{orange}977,958}$ MinOpts, ${\color{orange}1,517,282}$ FullOpts).

${\color{blue}MISSED}$ contexts: ${\color{blue}3,546 (0.14\\%)}$


<details>
<summary>Overall (${\color{green}-4,578}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|aspnet.run.windows.x64.checked.mch|42,070,891|${\color{green}-3,161}$|
|benchmarks.run.windows.x64.checked.mch|8,666,396|${\color{red}+2}$|
|benchmarks.run_pgo.windows.x64.checked.mch|35,349,995|${\color{green}-8,059}$|
|benchmarks.run_tiered.windows.x64.checked.mch|12,686,993|${\color{green}-10,190}$|
|coreclr_tests.run.windows.x64.checked.mch|393,136,663|${\color{red}+155,469}$|
|libraries.crossgen2.windows.x64.checked.mch|39,411,196|${\color{green}-2}$|
|libraries.pmi.windows.x64.checked.mch|61,131,115|${\color{red}+2}$|
|libraries_tests.run.windows.x64.Release.mch|279,572,690|${\color{green}-126,077}$|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|133,740,839|${\color{green}-12,568}$|
|realworld.run.windows.x64.checked.mch|14,188,290|${\color{red}+6}$|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,085,035|+0|


</div></details>

<details>
<summary>MinOpts (${\color{green}-4,578}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|aspnet.run.windows.x64.checked.mch|14,658,725|${\color{green}-3,161}$|
|benchmarks.run.windows.x64.checked.mch|361|${\color{red}+2}$|
|benchmarks.run_pgo.windows.x64.checked.mch|14,238,248|${\color{green}-8,059}$|
|benchmarks.run_tiered.windows.x64.checked.mch|9,174,699|${\color{green}-10,190}$|
|coreclr_tests.run.windows.x64.checked.mch|273,514,441|${\color{red}+155,469}$|
|libraries.crossgen2.windows.x64.checked.mch|1,189|${\color{green}-2}$|
|libraries.pmi.windows.x64.checked.mch|113,519|${\color{red}+2}$|
|libraries_tests.run.windows.x64.Release.mch|175,492,237|${\color{green}-126,077}$|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|10,810,850|${\color{green}-12,568}$|
|realworld.run.windows.x64.checked.mch|389,705|${\color{red}+6}$|
|smoke_tests.nativeaot.windows.x64.checked.mch|909|+0|


</div></details>


Details [here](https://clrjit2.blob.core.windows.net/diff-results/96386/COMMIT_HASH/2024-01-19_22-11-46-287/overall_asmdiffs_windows_x64.html)
### Assembly diffs for linux/arm ran on windows/x86
Diffs are based on ${\color{orange}2,232,466}$ contexts (${\color{orange}824,262}$ MinOpts, ${\color{orange}1,408,204}$ FullOpts).

${\color{blue}MISSED}$ contexts: ${\color{blue}77,883 (3.37\\%)}$


<details>
<summary>Overall (${\color{green}-852,700}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.linux.arm.checked.mch|14,172,892|${\color{green}-42}$|
|benchmarks.run_pgo.linux.arm.checked.mch|58,450,272|${\color{green}-48,252}$|
|benchmarks.run_tiered.linux.arm.checked.mch|17,364,030|${\color{green}-32,012}$|
|coreclr_tests.run.linux.arm.checked.mch|321,914,498|${\color{green}-199,178}$|
|libraries.crossgen2.linux.arm.checked.mch|37,768,318|+0|
|libraries.pmi.linux.arm.checked.mch|49,500,180|${\color{green}-280}$|
|libraries_tests.run.linux.arm.Release.mch|240,715,714|${\color{green}-533,380}$|
|libraries_tests_no_tiered_compilation.run.linux.arm.Release.mch|92,863,410|${\color{green}-24,910}$|
|realworld.run.linux.arm.checked.mch|13,608,080|${\color{green}-14,646}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-852,700}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.linux.arm.checked.mch|334,862|${\color{green}-42}$|
|benchmarks.run_pgo.linux.arm.checked.mch|11,222,492|${\color{green}-48,252}$|
|benchmarks.run_tiered.linux.arm.checked.mch|7,194,320|${\color{green}-32,012}$|
|coreclr_tests.run.linux.arm.checked.mch|212,840,102|${\color{green}-199,178}$|
|libraries.crossgen2.linux.arm.checked.mch|1,230|+0|
|libraries.pmi.linux.arm.checked.mch|106,504|${\color{green}-280}$|
|libraries_tests.run.linux.arm.Release.mch|122,210,132|${\color{green}-533,380}$|
|libraries_tests_no_tiered_compilation.run.linux.arm.Release.mch|10,081,872|${\color{green}-24,910}$|
|realworld.run.linux.arm.checked.mch|449,296|${\color{green}-14,646}$|


</div></details>

### Assembly diffs for windows/x86 ran on windows/x86
Diffs are based on ${\color{orange}2,240,380}$ contexts (${\color{orange}789,389}$ MinOpts, ${\color{orange}1,450,991}$ FullOpts).

${\color{blue}MISSED}$ contexts: base: ${\color{blue}3,847 (0.17\\%)}$, diff: ${\color{blue}49,158 (2.15\\%)}$


<details>
<summary>Overall (${\color{red}+199,265}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.windows.x86.checked.mch|7,037,944|${\color{red}+2}$|
|benchmarks.run_pgo.windows.x86.checked.mch|43,076,375|${\color{red}+1,814}$|
|benchmarks.run_tiered.windows.x86.checked.mch|8,947,008|${\color{red}+247}$|
|coreclr_tests.run.windows.x86.checked.mch|305,219,044|${\color{red}+201,221}$|
|libraries.crossgen2.windows.x86.checked.mch|31,623,120|${\color{red}+3}$|
|libraries.pmi.windows.x86.checked.mch|48,807,890|${\color{green}-81}$|
|libraries_tests.run.windows.x86.Release.mch|162,974,883|${\color{green}-7,140}$|
|libraries_tests_no_tiered_compilation.run.windows.x86.Release.mch|100,565,179|${\color{red}+3,185}$|
|realworld.run.windows.x86.checked.mch|11,349,106|${\color{red}+14}$|


</div></details>

<details>
<summary>MinOpts (${\color{red}+199,265}$ bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run.windows.x86.checked.mch|279|${\color{red}+2}$|
|benchmarks.run_pgo.windows.x86.checked.mch|5,966,425|${\color{red}+1,814}$|
|benchmarks.run_tiered.windows.x86.checked.mch|3,848,434|${\color{red}+247}$|
|coreclr_tests.run.windows.x86.checked.mch|198,156,671|${\color{red}+201,221}$|
|libraries.crossgen2.windows.x86.checked.mch|1,057|${\color{red}+3}$|
|libraries.pmi.windows.x86.checked.mch|95,314|${\color{green}-81}$|
|libraries_tests.run.windows.x86.Release.mch|77,656,212|${\color{green}-7,140}$|
|libraries_tests_no_tiered_compilation.run.windows.x86.Release.mch|7,273,446|${\color{red}+3,185}$|
|realworld.run.windows.x86.checked.mch|295,700|${\color{red}+14}$|


</div></details>


Details [here](https://clrjit2.blob.core.windows.net/diff-results/96386/COMMIT_HASH/2024-01-19_22-11-46-287/overall_asmdiffs_windows_x86.html)
## Throughput diffs

### Throughput diffs for linux/arm64 ran on windows/x64

<details>
<summary>Overall (${\color{green}-4.82\%}$ to ${\color{green}-0.04\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm64.checked.mch|${\color{green}-0.09\\%}$|
|benchmarks.run_pgo.linux.arm64.checked.mch|${\color{green}-1.12\\%}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|${\color{green}-4.82\\%}$|
|coreclr_tests.run.linux.arm64.checked.mch|${\color{green}-4.00\\%}$|
|libraries.crossgen2.linux.arm64.checked.mch|${\color{green}-0.07\\%}$|
|libraries.pmi.linux.arm64.checked.mch|${\color{green}-0.05\\%}$|
|libraries_tests.run.linux.arm64.Release.mch|${\color{green}-2.41\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|${\color{green}-0.30\\%}$|
|realworld.run.linux.arm64.checked.mch|${\color{green}-0.13\\%}$|
|smoke_tests.nativeaot.linux.arm64.checked.mch|${\color{green}-0.04\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-10.51\%}$ to ${\color{green}-6.50\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm64.checked.mch|${\color{green}-8.44\\%}$|
|benchmarks.run_pgo.linux.arm64.checked.mch|${\color{green}-9.02\\%}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|${\color{green}-9.30\\%}$|
|coreclr_tests.run.linux.arm64.checked.mch|${\color{green}-9.24\\%}$|
|libraries.crossgen2.linux.arm64.checked.mch|${\color{green}-8.09\\%}$|
|libraries.pmi.linux.arm64.checked.mch|${\color{green}-9.87\\%}$|
|libraries_tests.run.linux.arm64.Release.mch|${\color{green}-9.06\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|${\color{green}-8.69\\%}$|
|realworld.run.linux.arm64.checked.mch|${\color{green}-10.51\\%}$|
|smoke_tests.nativeaot.linux.arm64.checked.mch|${\color{green}-6.50\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.07\%}$ to ${\color{green}-0.03\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm64.checked.mch|${\color{green}-0.05\\%}$|
|benchmarks.run_pgo.linux.arm64.checked.mch|${\color{green}-0.04\\%}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|${\color{green}-0.03\\%}$|
|coreclr_tests.run.linux.arm64.checked.mch|${\color{green}-0.05\\%}$|
|libraries.crossgen2.linux.arm64.checked.mch|${\color{green}-0.07\\%}$|
|libraries.pmi.linux.arm64.checked.mch|${\color{green}-0.04\\%}$|
|libraries_tests.run.linux.arm64.Release.mch|${\color{green}-0.04\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|${\color{green}-0.05\\%}$|
|realworld.run.linux.arm64.checked.mch|${\color{green}-0.04\\%}$|
|smoke_tests.nativeaot.linux.arm64.checked.mch|${\color{green}-0.04\\%}$|


</div></details>

### Throughput diffs for linux/x64 ran on windows/x64

<details>
<summary>Overall (${\color{green}-2.72\%}$ to ${\color{green}-0.05\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.x64.checked.mch|${\color{green}-0.09\\%}$|
|benchmarks.run_pgo.linux.x64.checked.mch|${\color{green}-0.56\\%}$|
|benchmarks.run_tiered.linux.x64.checked.mch|${\color{green}-2.72\\%}$|
|coreclr_tests.run.linux.x64.checked.mch|${\color{green}-2.70\\%}$|
|libraries.crossgen2.linux.x64.checked.mch|${\color{green}-0.08\\%}$|
|libraries.pmi.linux.x64.checked.mch|${\color{green}-0.06\\%}$|
|libraries_tests.run.linux.x64.Release.mch|${\color{green}-1.34\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|${\color{green}-0.21\\%}$|
|realworld.run.linux.x64.checked.mch|${\color{green}-0.10\\%}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|${\color{green}-0.05\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-6.56\%}$ to ${\color{green}-4.49\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.x64.checked.mch|${\color{green}-5.03\\%}$|
|benchmarks.run_pgo.linux.x64.checked.mch|${\color{green}-5.63\\%}$|
|benchmarks.run_tiered.linux.x64.checked.mch|${\color{green}-5.71\\%}$|
|coreclr_tests.run.linux.x64.checked.mch|${\color{green}-6.24\\%}$|
|libraries.crossgen2.linux.x64.checked.mch|${\color{green}-4.49\\%}$|
|libraries.pmi.linux.x64.checked.mch|${\color{green}-5.88\\%}$|
|libraries_tests.run.linux.x64.Release.mch|${\color{green}-5.66\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|${\color{green}-5.15\\%}$|
|realworld.run.linux.x64.checked.mch|${\color{green}-6.56\\%}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|${\color{green}-4.77\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.08\%}$ to ${\color{green}-0.04\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.x64.checked.mch|${\color{green}-0.06\\%}$|
|benchmarks.run_pgo.linux.x64.checked.mch|${\color{green}-0.04\\%}$|
|benchmarks.run_tiered.linux.x64.checked.mch|${\color{green}-0.04\\%}$|
|coreclr_tests.run.linux.x64.checked.mch|${\color{green}-0.06\\%}$|
|libraries.crossgen2.linux.x64.checked.mch|${\color{green}-0.08\\%}$|
|libraries.pmi.linux.x64.checked.mch|${\color{green}-0.06\\%}$|
|libraries_tests.run.linux.x64.Release.mch|${\color{green}-0.05\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|${\color{green}-0.07\\%}$|
|realworld.run.linux.x64.checked.mch|${\color{green}-0.06\\%}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|${\color{green}-0.05\\%}$|


</div></details>

### Throughput diffs for osx/arm64 ran on windows/x64

<details>
<summary>Overall (${\color{green}-4.32\%}$ to ${\color{green}-0.04\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.osx.arm64.checked.mch|${\color{green}-0.04\\%}$|
|benchmarks.run_pgo.osx.arm64.checked.mch|${\color{green}-1.94\\%}$|
|benchmarks.run_tiered.osx.arm64.checked.mch|${\color{green}-4.32\\%}$|
|coreclr_tests.run.osx.arm64.checked.mch|${\color{green}-3.96\\%}$|
|libraries.crossgen2.osx.arm64.checked.mch|${\color{green}-0.07\\%}$|
|libraries.pmi.osx.arm64.checked.mch|${\color{green}-0.05\\%}$|
|libraries_tests.run.osx.arm64.Release.mch|${\color{green}-3.05\\%}$|
|libraries_tests_no_tiered_compilation.run.osx.arm64.Release.mch|${\color{green}-0.30\\%}$|
|realworld.run.osx.arm64.checked.mch|${\color{green}-0.13\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-10.55\%}$ to ${\color{green}-7.99\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.osx.arm64.checked.mch|${\color{green}-9.05\\%}$|
|benchmarks.run_pgo.osx.arm64.checked.mch|${\color{green}-9.34\\%}$|
|benchmarks.run_tiered.osx.arm64.checked.mch|${\color{green}-9.62\\%}$|
|coreclr_tests.run.osx.arm64.checked.mch|${\color{green}-9.13\\%}$|
|libraries.crossgen2.osx.arm64.checked.mch|${\color{green}-7.99\\%}$|
|libraries.pmi.osx.arm64.checked.mch|${\color{green}-9.86\\%}$|
|libraries_tests.run.osx.arm64.Release.mch|${\color{green}-9.09\\%}$|
|libraries_tests_no_tiered_compilation.run.osx.arm64.Release.mch|${\color{green}-8.68\\%}$|
|realworld.run.osx.arm64.checked.mch|${\color{green}-10.55\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.07\%}$ to ${\color{green}-0.03\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.osx.arm64.checked.mch|${\color{green}-0.04\\%}$|
|benchmarks.run_pgo.osx.arm64.checked.mch|${\color{green}-0.03\\%}$|
|benchmarks.run_tiered.osx.arm64.checked.mch|${\color{green}-0.03\\%}$|
|coreclr_tests.run.osx.arm64.checked.mch|${\color{green}-0.06\\%}$|
|libraries.crossgen2.osx.arm64.checked.mch|${\color{green}-0.07\\%}$|
|libraries.pmi.osx.arm64.checked.mch|${\color{green}-0.04\\%}$|
|libraries_tests.run.osx.arm64.Release.mch|${\color{green}-0.04\\%}$|
|libraries_tests_no_tiered_compilation.run.osx.arm64.Release.mch|${\color{green}-0.05\\%}$|
|realworld.run.osx.arm64.checked.mch|${\color{green}-0.04\\%}$|


</div></details>

### Throughput diffs for windows/arm64 ran on windows/x64

<details>
<summary>Overall (${\color{green}-4.16\%}$ to ${\color{green}-0.04\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.arm64.checked.mch|${\color{green}-0.04\\%}$|
|benchmarks.run_pgo.windows.arm64.checked.mch|${\color{green}-1.27\\%}$|
|benchmarks.run_tiered.windows.arm64.checked.mch|${\color{green}-4.16\\%}$|
|coreclr_tests.run.windows.arm64.checked.mch|${\color{green}-4.01\\%}$|
|libraries.crossgen2.windows.arm64.checked.mch|${\color{green}-0.07\\%}$|
|libraries.pmi.windows.arm64.checked.mch|${\color{green}-0.05\\%}$|
|libraries_tests.run.windows.arm64.Release.mch|${\color{green}-2.98\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.arm64.Release.mch|${\color{green}-0.28\\%}$|
|realworld.run.windows.arm64.checked.mch|${\color{green}-0.13\\%}$|
|smoke_tests.nativeaot.windows.arm64.checked.mch|${\color{green}-0.04\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-10.55\%}$ to ${\color{green}-6.32\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.arm64.checked.mch|${\color{green}-9.05\\%}$|
|benchmarks.run_pgo.windows.arm64.checked.mch|${\color{green}-9.25\\%}$|
|benchmarks.run_tiered.windows.arm64.checked.mch|${\color{green}-9.57\\%}$|
|coreclr_tests.run.windows.arm64.checked.mch|${\color{green}-9.25\\%}$|
|libraries.crossgen2.windows.arm64.checked.mch|${\color{green}-8.09\\%}$|
|libraries.pmi.windows.arm64.checked.mch|${\color{green}-9.86\\%}$|
|libraries_tests.run.windows.arm64.Release.mch|${\color{green}-9.07\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.arm64.Release.mch|${\color{green}-8.68\\%}$|
|realworld.run.windows.arm64.checked.mch|${\color{green}-10.55\\%}$|
|smoke_tests.nativeaot.windows.arm64.checked.mch|${\color{green}-6.32\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.07\%}$ to ${\color{green}-0.03\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.arm64.checked.mch|${\color{green}-0.04\\%}$|
|benchmarks.run_pgo.windows.arm64.checked.mch|${\color{green}-0.03\\%}$|
|benchmarks.run_tiered.windows.arm64.checked.mch|${\color{green}-0.03\\%}$|
|coreclr_tests.run.windows.arm64.checked.mch|${\color{green}-0.06\\%}$|
|libraries.crossgen2.windows.arm64.checked.mch|${\color{green}-0.07\\%}$|
|libraries.pmi.windows.arm64.checked.mch|${\color{green}-0.04\\%}$|
|libraries_tests.run.windows.arm64.Release.mch|${\color{green}-0.04\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.arm64.Release.mch|${\color{green}-0.05\\%}$|
|realworld.run.windows.arm64.checked.mch|${\color{green}-0.04\\%}$|
|smoke_tests.nativeaot.windows.arm64.checked.mch|${\color{green}-0.04\\%}$|


</div></details>

### Throughput diffs for windows/x64 ran on windows/x64

<details>
<summary>Overall (${\color{green}-2.54\%}$ to ${\color{green}-0.08\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|${\color{green}-0.66\\%}$|
|benchmarks.run.windows.x64.checked.mch|${\color{green}-0.08\\%}$|
|benchmarks.run_pgo.windows.x64.checked.mch|${\color{green}-0.83\\%}$|
|benchmarks.run_tiered.windows.x64.checked.mch|${\color{green}-2.22\\%}$|
|coreclr_tests.run.windows.x64.checked.mch|${\color{green}-2.54\\%}$|
|libraries.crossgen2.windows.x64.checked.mch|${\color{green}-0.12\\%}$|
|libraries.pmi.windows.x64.checked.mch|${\color{green}-0.10\\%}$|
|libraries_tests.run.windows.x64.Release.mch|${\color{green}-1.64\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|${\color{green}-0.25\\%}$|
|realworld.run.windows.x64.checked.mch|${\color{green}-0.13\\%}$|
|smoke_tests.nativeaot.windows.x64.checked.mch|${\color{green}-0.08\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-6.49\%}$ to ${\color{green}-4.45\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|${\color{green}-5.50\\%}$|
|benchmarks.run.windows.x64.checked.mch|${\color{green}-4.89\\%}$|
|benchmarks.run_pgo.windows.x64.checked.mch|${\color{green}-5.70\\%}$|
|benchmarks.run_tiered.windows.x64.checked.mch|${\color{green}-5.72\\%}$|
|coreclr_tests.run.windows.x64.checked.mch|${\color{green}-5.95\\%}$|
|libraries.crossgen2.windows.x64.checked.mch|${\color{green}-4.45\\%}$|
|libraries.pmi.windows.x64.checked.mch|${\color{green}-5.72\\%}$|
|libraries_tests.run.windows.x64.Release.mch|${\color{green}-5.74\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|${\color{green}-5.52\\%}$|
|realworld.run.windows.x64.checked.mch|${\color{green}-6.49\\%}$|
|smoke_tests.nativeaot.windows.x64.checked.mch|${\color{green}-4.70\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.12\%}$ to ${\color{green}-0.07\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|${\color{green}-0.07\\%}$|
|benchmarks.run.windows.x64.checked.mch|${\color{green}-0.08\\%}$|
|benchmarks.run_pgo.windows.x64.checked.mch|${\color{green}-0.07\\%}$|
|benchmarks.run_tiered.windows.x64.checked.mch|${\color{green}-0.07\\%}$|
|coreclr_tests.run.windows.x64.checked.mch|${\color{green}-0.11\\%}$|
|libraries.crossgen2.windows.x64.checked.mch|${\color{green}-0.12\\%}$|
|libraries.pmi.windows.x64.checked.mch|${\color{green}-0.09\\%}$|
|libraries_tests.run.windows.x64.Release.mch|${\color{green}-0.08\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|${\color{green}-0.11\\%}$|
|realworld.run.windows.x64.checked.mch|${\color{green}-0.10\\%}$|
|smoke_tests.nativeaot.windows.x64.checked.mch|${\color{green}-0.08\\%}$|


</div></details>


Details [here](https://clrjit2.blob.core.windows.net/diff-results/96386/COMMIT_HASH/2024-01-19_22-11-46-287/overall_tpdiffs_windows_x64.html)
### Throughput diffs for linux/arm64 ran on linux/x64

<details>
<summary>Overall (${\color{green}-5.32\%}$ to ${\color{red}+0.04\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|libraries.crossgen2.linux.arm64.checked.mch|${\color{red}+0.04\\%}$|
|libraries_tests.run.linux.arm64.Release.mch|${\color{green}-2.61\\%}$|
|realworld.run.linux.arm64.checked.mch|${\color{green}-0.07\\%}$|
|coreclr_tests.run.linux.arm64.checked.mch|${\color{green}-4.33\\%}$|
|benchmarks.run_pgo.linux.arm64.checked.mch|${\color{green}-1.17\\%}$|
|libraries.pmi.linux.arm64.checked.mch|${\color{red}+0.03\\%}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|${\color{green}-5.32\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|${\color{green}-0.24\\%}$|
|smoke_tests.nativeaot.linux.arm64.checked.mch|${\color{red}+0.04\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-11.63\%}$ to ${\color{green}-6.46\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|libraries.crossgen2.linux.arm64.checked.mch|${\color{green}-9.09\\%}$|
|libraries_tests.run.linux.arm64.Release.mch|${\color{green}-9.98\\%}$|
|realworld.run.linux.arm64.checked.mch|${\color{green}-11.63\\%}$|
|coreclr_tests.run.linux.arm64.checked.mch|${\color{green}-9.71\\%}$|
|benchmarks.run_pgo.linux.arm64.checked.mch|${\color{green}-9.97\\%}$|
|libraries.pmi.linux.arm64.checked.mch|${\color{green}-11.20\\%}$|
|benchmarks.run.linux.arm64.checked.mch|${\color{green}-9.20\\%}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|${\color{green}-10.35\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|${\color{green}-10.06\\%}$|
|smoke_tests.nativeaot.linux.arm64.checked.mch|${\color{green}-6.46\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{red}+0.02\%}$ to ${\color{red}+0.05\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|libraries.crossgen2.linux.arm64.checked.mch|${\color{red}+0.04\\%}$|
|libraries_tests.run.linux.arm64.Release.mch|${\color{red}+0.03\\%}$|
|realworld.run.linux.arm64.checked.mch|${\color{red}+0.03\\%}$|
|coreclr_tests.run.linux.arm64.checked.mch|${\color{red}+0.05\\%}$|
|benchmarks.run_pgo.linux.arm64.checked.mch|${\color{red}+0.03\\%}$|
|libraries.pmi.linux.arm64.checked.mch|${\color{red}+0.03\\%}$|
|benchmarks.run.linux.arm64.checked.mch|${\color{red}+0.05\\%}$|
|benchmarks.run_tiered.linux.arm64.checked.mch|${\color{red}+0.02\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm64.Release.mch|${\color{red}+0.05\\%}$|
|smoke_tests.nativeaot.linux.arm64.checked.mch|${\color{red}+0.04\\%}$|


</div></details>

### Throughput diffs for linux/x64 ran on linux/x64

<details>
<summary>Overall (${\color{green}-2.87\%}$ to ${\color{green}-0.15\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run_pgo.linux.x64.checked.mch|${\color{green}-0.66\\%}$|
|benchmarks.run_tiered.linux.x64.checked.mch|${\color{green}-2.87\\%}$|
|libraries.crossgen2.linux.x64.checked.mch|${\color{green}-0.22\\%}$|
|libraries.pmi.linux.x64.checked.mch|${\color{green}-0.17\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|${\color{green}-0.34\\%}$|
|realworld.run.linux.x64.checked.mch|${\color{green}-0.20\\%}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|${\color{green}-0.15\\%}$|
|benchmarks.run.linux.x64.checked.mch|${\color{green}-0.20\\%}$|
|libraries_tests.run.linux.x64.Release.mch|${\color{green}-1.45\\%}$|
|coreclr_tests.run.linux.x64.checked.mch|${\color{green}-2.80\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-6.69\%}$ to ${\color{green}-4.65\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run_pgo.linux.x64.checked.mch|${\color{green}-5.91\\%}$|
|benchmarks.run_tiered.linux.x64.checked.mch|${\color{green}-5.96\\%}$|
|libraries.crossgen2.linux.x64.checked.mch|${\color{green}-4.86\\%}$|
|libraries.pmi.linux.x64.checked.mch|${\color{green}-6.32\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|${\color{green}-5.60\\%}$|
|realworld.run.linux.x64.checked.mch|${\color{green}-6.69\\%}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|${\color{green}-4.65\\%}$|
|benchmarks.run.linux.x64.checked.mch|${\color{green}-5.12\\%}$|
|libraries_tests.run.linux.x64.Release.mch|${\color{green}-5.81\\%}$|
|coreclr_tests.run.linux.x64.checked.mch|${\color{green}-6.04\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.22\%}$ to ${\color{green}-0.12\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run_pgo.linux.x64.checked.mch|${\color{green}-0.13\\%}$|
|benchmarks.run_tiered.linux.x64.checked.mch|${\color{green}-0.12\\%}$|
|libraries.crossgen2.linux.x64.checked.mch|${\color{green}-0.22\\%}$|
|libraries.pmi.linux.x64.checked.mch|${\color{green}-0.16\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.x64.Release.mch|${\color{green}-0.19\\%}$|
|realworld.run.linux.x64.checked.mch|${\color{green}-0.16\\%}$|
|smoke_tests.nativeaot.linux.x64.checked.mch|${\color{green}-0.15\\%}$|
|benchmarks.run.linux.x64.checked.mch|${\color{green}-0.17\\%}$|
|libraries_tests.run.linux.x64.Release.mch|${\color{green}-0.14\\%}$|
|coreclr_tests.run.linux.x64.checked.mch|${\color{green}-0.18\\%}$|


</div></details>


Details [here](https://clrjit2.blob.core.windows.net/diff-results/96386/COMMIT_HASH/2024-01-19_22-11-46-287/overall_tpdiffs_linux_x64.html)
### Throughput diffs for linux/arm ran on windows/x86

<details>
<summary>Overall (${\color{green}-4.15\%}$ to ${\color{green}-0.08\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm.checked.mch|${\color{green}-0.18\\%}$|
|benchmarks.run_pgo.linux.arm.checked.mch|${\color{green}-0.72\\%}$|
|benchmarks.run_tiered.linux.arm.checked.mch|${\color{green}-2.03\\%}$|
|coreclr_tests.run.linux.arm.checked.mch|${\color{green}-4.15\\%}$|
|libraries.crossgen2.linux.arm.checked.mch|${\color{green}-0.15\\%}$|
|libraries.pmi.linux.arm.checked.mch|${\color{green}-0.08\\%}$|
|libraries_tests.run.linux.arm.Release.mch|${\color{green}-2.51\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm.Release.mch|${\color{green}-0.47\\%}$|
|realworld.run.linux.arm.checked.mch|${\color{green}-0.15\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-11.56\%}$ to ${\color{green}-9.30\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm.checked.mch|${\color{green}-9.64\\%}$|
|benchmarks.run_pgo.linux.arm.checked.mch|${\color{green}-10.51\\%}$|
|benchmarks.run_tiered.linux.arm.checked.mch|${\color{green}-10.52\\%}$|
|coreclr_tests.run.linux.arm.checked.mch|${\color{green}-9.88\\%}$|
|libraries.crossgen2.linux.arm.checked.mch|${\color{green}-9.83\\%}$|
|libraries.pmi.linux.arm.checked.mch|${\color{green}-9.30\\%}$|
|libraries_tests.run.linux.arm.Release.mch|${\color{green}-10.61\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm.Release.mch|${\color{green}-9.92\\%}$|
|realworld.run.linux.arm.checked.mch|${\color{green}-11.56\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{green}-0.15\%}$ to ${\color{green}-0.05\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm.checked.mch|${\color{green}-0.09\\%}$|
|benchmarks.run_pgo.linux.arm.checked.mch|${\color{green}-0.05\\%}$|
|benchmarks.run_tiered.linux.arm.checked.mch|${\color{green}-0.09\\%}$|
|coreclr_tests.run.linux.arm.checked.mch|${\color{green}-0.11\\%}$|
|libraries.crossgen2.linux.arm.checked.mch|${\color{green}-0.15\\%}$|
|libraries.pmi.linux.arm.checked.mch|${\color{green}-0.07\\%}$|
|libraries_tests.run.linux.arm.Release.mch|${\color{green}-0.05\\%}$|
|libraries_tests_no_tiered_compilation.run.linux.arm.Release.mch|${\color{green}-0.09\\%}$|
|realworld.run.linux.arm.checked.mch|${\color{green}-0.06\\%}$|


</div></details>

### Throughput diffs for windows/x86 ran on windows/x86

<details>
<summary>Overall (${\color{green}-1.68\%}$ to ${\color{red}+0.02\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x86.checked.mch|${\color{red}+0.02\\%}$|
|benchmarks.run_pgo.windows.x86.checked.mch|${\color{green}-0.12\\%}$|
|benchmarks.run_tiered.windows.x86.checked.mch|${\color{green}-0.59\\%}$|
|coreclr_tests.run.windows.x86.checked.mch|${\color{green}-1.68\\%}$|
|libraries.crossgen2.windows.x86.checked.mch|${\color{red}+0.01\\%}$|
|libraries.pmi.windows.x86.checked.mch|${\color{red}+0.02\\%}$|
|libraries_tests.run.windows.x86.Release.mch|${\color{green}-0.74\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.x86.Release.mch|${\color{green}-0.07\\%}$|
|realworld.run.windows.x86.checked.mch|${\color{green}-0.01\\%}$|


</div></details>

<details>
<summary>MinOpts (${\color{green}-5.03\%}$ to ${\color{green}-3.05\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x86.checked.mch|${\color{green}-3.15\\%}$|
|benchmarks.run_pgo.windows.x86.checked.mch|${\color{green}-3.99\\%}$|
|benchmarks.run_tiered.windows.x86.checked.mch|${\color{green}-3.90\\%}$|
|coreclr_tests.run.windows.x86.checked.mch|${\color{green}-5.03\\%}$|
|libraries.crossgen2.windows.x86.checked.mch|${\color{green}-3.05\\%}$|
|libraries.pmi.windows.x86.checked.mch|${\color{green}-4.50\\%}$|
|libraries_tests.run.windows.x86.Release.mch|${\color{green}-4.24\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.x86.Release.mch|${\color{green}-3.69\\%}$|
|realworld.run.windows.x86.checked.mch|${\color{green}-4.85\\%}$|


</div></details>

<details>
<summary>FullOpts (${\color{red}+0.01\%}$ to ${\color{red}+0.02\%}$)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x86.checked.mch|${\color{red}+0.02\\%}$|
|benchmarks.run_pgo.windows.x86.checked.mch|${\color{red}+0.02\\%}$|
|benchmarks.run_tiered.windows.x86.checked.mch|${\color{red}+0.02\\%}$|
|coreclr_tests.run.windows.x86.checked.mch|${\color{red}+0.01\\%}$|
|libraries.crossgen2.windows.x86.checked.mch|${\color{red}+0.01\\%}$|
|libraries.pmi.windows.x86.checked.mch|${\color{red}+0.02\\%}$|
|libraries_tests.run.windows.x86.Release.mch|${\color{red}+0.02\\%}$|
|libraries_tests_no_tiered_compilation.run.windows.x86.Release.mch|${\color{red}+0.01\\%}$|
|realworld.run.windows.x86.checked.mch|${\color{red}+0.02\\%}$|


</div></details>


Details [here](https://clrjit2.blob.core.windows.net/diff-results/96386/COMMIT_HASH/2024-01-19_22-11-46-287/overall_tpdiffs_windows_x86.html)

kunalspathak commented:
@BruceForstall @jakobbotsch - can you please take another look? Thanks!

jakobbotsch commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -3204,6 +3260,53 @@ bool LinearScan::isSpillCandidate(Interval* current, RefPosition* refPosition, R
     return canSpill;
 }
 
+// Grab a register to use to copy and then immediately use.
+// This is called only for localVar intervals that already have a register
+// assignment that is not compatible with the current RefPosition.
+// This is not like regular assignment, because we don't want to change
+// any preferences or existing register assignments.
+// Prefer a free register that's got the earliest next use.
+// Otherwise, spill something with the farthest next use
+//
+regNumber LinearScan::assignCopyRegMinimal(RefPosition* refPosition)
```
Nit: normal function header format?

jakobbotsch commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -4725,9 +4835,746 @@ void LinearScan::freeRegisters(regMaskTP regsToFree)
+                regsToFree      = delayRegsToFree;
+                delayRegsToFree = RBM_NONE;
+
+#ifdef DEBUG
+                // Validate the current state just after we've freed the registers. This ensures that any pending
+                // freed registers will have had their state updated to reflect the intervals they were holding.
+                for (regNumber reg = REG_FIRST; reg < AVAILABLE_REG_COUNT; reg = REG_NEXT(reg))
```
Is it possible/straightforward to factor this validation into a method and use it from both `allocateRegisters` and `allocateRegistersMinimal`?

A related question -- what are the fundamental difference between the two, except the reduced set of heuristics used in the minimal version?

jakobbotsch reviewed (approved):
I've skimmed the code. I'm assuming `allocateRegistersMinimal` was copied mostly from `allocateRegisters`. I had a question on what the differences between them is, apart from the use of a reduced set of heuristics -- I'm assuming it wouldn't be straightforward to unify them to avoid code duplication? (For example, by the use of another template parameter)

The trade offs in CQ vs TP look just fine to me. We should definitely be taking this kind of tradeoff in MinOpts, and the TP improvements are amazing.

kunalspathak commented:
>  I'm assuming allocateRegistersMinimal was copied mostly from allocateRegisters.

Pretty much. To easily spot the difference, I extracted out here:

- [allocateregsMinimal](https://www.diffchecker.com/iwKkJeO9/)
- [assigncopyRegMinimal](https://www.diffchecker.com/opAnpzQK/)
- [allocateRegistersMinimal](https://www.diffchecker.com/0RM6a4pD/)
- [selectMinimal](https://www.diffchecker.com/OvsP3toN/)
- [resetMinimal](https://www.diffchecker.com/jo0VddT3/)

>  I'm assuming it wouldn't be straightforward to unify them to avoid code duplication? (For example, by the use of another template parameter)

Correct, that was the direction I started with, but I quickly noticed that it scatters the templated parameter everywhere and very hard to see what gets executed for MinOpts. In future, it will be good to focus on `*minimal()` APIs to further simplify the algorithm.

kunalspathak committed:
Add verifyFreeRegisters()

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into minopts-lsra
